In [5]:
# Import libraries
import arff
import pandas as pd
import numpy as np

In [6]:
# Load the arff file
with open('titanic_data.arff', 'r') as f:
    dataset = arff.load(f)

In [7]:
# Load the data and column names
data = dataset["data"]
column_names = [attr[0] for attr in dataset["attributes"]]

In [8]:
# Create the DataFrame
df = pd.DataFrame(data, columns=column_names)
df.head(10)

pclass survived                                             name     sex  \
0     1.0        1                    Allen, Miss. Elisabeth Walton  female   
1     1.0        1                   Allison, Master. Hudson Trevor    male   
2     1.0        0                     Allison, Miss. Helen Loraine  female   
3     1.0        0             Allison, Mr. Hudson Joshua Creighton    male   
4     1.0        0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   
5     1.0        1                              Anderson, Mr. Harry    male   
6     1.0        1                Andrews, Miss. Kornelia Theodosia  female   
7     1.0        0                           Andrews, Mr. Thomas Jr    male   
8     1.0        1    Appleton, Mrs. Edward Dale (Charlotte Lamson)  female   
9     1.0        0                          Artagaveytia, Mr. Ramon    male   

       age  sibsp  parch    ticket      fare    cabin embarked  boat   body  \
0  29.0000    0.0    0.0     24160  211.3375       B5        S     2    NaN   
1   0.9167    1.0    2.0    113781  151.5500  C22 C26        S    11    NaN   
2   2.0000    1.0    2.0    113781  151.5500  C22 C26        S  None    NaN   
3  30.0000    1.0    2.0    113781  151.5500  C22 C26        S  None  135.0   
4  25.0000    1.0    2.0    113781  151.5500  C22 C26        S  None    NaN   
5  48.0000    0.0    0.0     19952   26.5500      E12        S     3    NaN   
6  63.0000    1.0    0.0     13502   77.9583       D7        S    10    NaN   
7  39.0000    0.0    0.0    112050    0.0000      A36        S  None    NaN   
8  53.0000    2.0    0.0     11769   51.4792     C101        S     D    NaN   
9  71.0000    0.0    0.0  PC 17609   49.5042     None        C  None   22.0   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON  
5                     New York, NY  
6                       Hudson, NY  
7                      Belfast, NI  
8              Bayside, Queens, NY  
9              Montevideo, Uruguay

Qualitative Data is also called Categorical Data and it categorizes the data into various categories.

Qualitative data is further categorized into two categories that include:
- Nominal Data - a type of data that consists of categories or names that cannot be ordered or ranked (e.g. gender, race, religion, blood type)
- Ordinal Data - a type of data that consists of categories that can be ordered or ranked (e.g. job position, education level)

Source: https://www.geeksforgeeks.org/data-types-in-statistics/#qualitative-data-categorical-data


In [9]:
# Check the number of individual labels for given qualitative variables

qual_features = [
    'pclass',
    'name',
    'sex',
    'ticket',
    'cabin',
    'embarked',
    'boat',
    'home.dest'
]

qual_df = df[qual_features]

for feature in qual_df.columns:
    print(f'No of labels of the "{feature}" feature: {len(df[feature].unique())}.')

No of labels of the "pclass" feature: 3.
No of labels of the "name" feature: 1307.
No of labels of the "sex" feature: 2.
No of labels of the "ticket" feature: 929.
No of labels of the "cabin" feature: 187.
No of labels of the "embarked" feature: 4.
No of labels of the "boat" feature: 28.
No of labels of the "home.dest" feature: 370.


In [6]:
# Chceck the number of all passengers

print('The number of all passengers is: {}.'.format(len(data)))

The number of all passengers is: 1309.


As we can see, there are features with many different labels. The example is the 'ticket' feature, because almost every passenger paid a different price for the ticket. Also features with a small number of lables appear in the data. There were only three passenger's ticket classes, so 'pclass' can take only three different values.

Cardinality refers to the number of unique values in a dataset column.
A column with high cardinality contains a vast number of unique values, while columns with low cardinality contain fewer unique entries.
(source: https://www.datacamp.com/tutorial/cardinality)

We can divide the features into two categories - these with low cardinality and high cardinality.

Low cardinality: 
- pclass
- survived
- sex
- sibps
- parch
- embarked
- boat

High cardinality: 
- name
- age
- ticket
- fare
- cabin
- body
- home.dest

I could not find a specific border to divide the features, so I based on my current knowledge and decide to group them in the way presented above.

In [7]:
# Check how many unique values the 'cabin' feature has
# Use a function that return a NumPy array

print(f'Type of the output: {type(df['cabin'].unique())}')
print(f'The number of unique values in the "cabin" column: {len(df['cabin'].unique())}')

Type of the output: <class 'numpy.ndarray'>
The number of unique values in the "cabin" column: 187


In [8]:
# Check the labels of the 'cabin' columns
print(df['cabin'].unique())

['B5' 'C22 C26' 'E12' 'D7' 'A36' 'C101' None 'C62 C64' 'B35' 'A23'
 'B58 B60' 'D15' 'C6' 'D35' 'C148' 'C97' 'B49' 'C99' 'C52' 'T' 'A31' 'C7'
 'C103' 'D22' 'E33' 'A21' 'B10' 'B4' 'E40' 'B38' 'E24' 'B51 B53 B55'
 'B96 B98' 'C46' 'E31' 'E8' 'B61' 'B77' 'A9' 'C89' 'A14' 'E58' 'E49' 'E52'
 'E45' 'B22' 'B26' 'C85' 'E17' 'B71' 'B20' 'A34' 'C86' 'A16' 'A20' 'A18'
 'C54' 'C45' 'D20' 'A29' 'C95' 'E25' 'C111' 'C23 C25 C27' 'E36' 'D34'
 'D40' 'B39' 'B41' 'B102' 'C123' 'E63' 'C130' 'B86' 'C92' 'A5' 'C51' 'B42'
 'C91' 'C125' 'D10 D12' 'B82 B84' 'E50' 'D33' 'C83' 'B94' 'D49' 'D45'
 'B69' 'B11' 'E46' 'C39' 'B18' 'D11' 'C93' 'B28' 'C49' 'B52 B54 B56' 'E60'
 'C132' 'B37' 'D21' 'D19' 'C124' 'D17' 'B101' 'D28' 'D6' 'D9' 'B80' 'C106'
 'B79' 'C47' 'D30' 'C90' 'E38' 'C78' 'C30' 'C118' 'D36' 'D48' 'D47' 'C105'
 'B36' 'B30' 'D43' 'B24' 'C2' 'C65' 'B73' 'C104' 'C110' 'C50' 'B3' 'A24'
 'A32' 'A11' 'A10' 'B57 B59 B63 B66' 'C28' 'E44' 'A26' 'A6' 'A7' 'C31'
 'A19' 'B45' 'E34' 'B78' 'B50' 'C87' 'C116' 'C55 C57' 'D50

In [9]:
# Reduce the number of labels in the 'cabin' columns adnd create a new column
# Replace the current format ('L11') with the label containing only the first letter ('L')

df['CabinReduced'] = df['cabin'].astype(str).str[0]

# Print the first 20 rows
df[['cabin', 'CabinReduced']].head(20)
print(sorted(df['CabinReduced'].unique()))

['A', 'B', 'C', 'D', 'E', 'F', 'G', 'N', 'T']


In [10]:
# Print the number of labels in the 'CabinReduced' column

print(f'No of labels of the "CabinReduced" feature: {len(df['CabinReduced'].unique())}')

No of labels of the "CabinReduced" feature: 9


In [11]:
# Show by what percentage the cardinality of the set of variables describing the 'cabin' column was reduced

original = len(df['cabin'].unique())
reduced = len(df['CabinReduced'].unique())

difference = 100 - (reduced * 100 / original)
print(f'The cardinality of the cabin feature has been reduced by {difference:.2f}%')

The cardinality of the cabin feature has been reduced by 95.19%


I decided to reduce the cardinality of the 'cabin' column, because by simply replacing the whole cabin name I decrase the number of labels from 187 to 9. I think the entire cabin name is not necessary in my analysis and having only a letter (which means the deck on the Titanic ship - https://www.encyclopedia-titanica.org/titanic-deckplans/) is enough to train the future model.

Reducing the cardinality positively influences the analysis. One of the biggest challenges of high cardinality is its impact on data storage and system performance. Also, it helps with data visualization, because high-cardinality data can hide insights instead of revealing them and make the plot unreadable. 

The more data and labels the more accurate model can be, however, by reducing the cardinality I can avoid the overfitting problem in my future model.


In [12]:
# Save the df as a pickle file
df.to_pickle('df_reduced.pkl')